Snackbar when mark read/unread from library

nzoba opened the pull request:
This PR contains:

- Normalize manga titles apostrophes to always be `U+0027` instead of `U-2019` (because most used and easiest to type with a keyboard), maybe would be better in tachiyomi-extensions instead ?
- Snackbar when mark manga as read/unread from library to allow undo (Closes #1116)
- Add language to pinned sources and migration pages &rarr; ex: `Mangadex (EN)` when multiples languages enabled (Closes #1091)
- Fix old chapters being readded in updates (taken from original Tachiyomi)
- Reader WebView opens chapter page instead of manga page (while WebView in manga_details still opens manga page)


| Undo read/unread | Language in pinned sources | Language in migration |
|----|----|----|
| ![Undo read/unread](https://user-images.githubusercontent.com/55888232/155901135-26798fd1-942c-4f3e-bb1d-2d1944aa076b.png) | ![Language in pinned sources](https://user-images.githubusercontent.com/55888232/155900060-11511b85-c6c4-4daf-8c8d-02c142448f77.png)  |  ![Language in migration](https://user-images.githubusercontent.com/55888232/155900081-b7481804-a33c-4cb4-a373-606da965f3e0.png)  |

Jays2Kings commented on app/src/main/java/eu/kanade/tachiyomi/ui/reader/ReaderActivity.kt:
```diff
@@ -1526,11 +1527,13 @@ class ReaderActivity : BaseRxActivity<ReaderPresenter>() {
     private fun openMangaInBrowser() {
         val source = presenter.getSource() ?: return
-        val url = try {
+        val chapterUrl = presenter.getCurrentChapter()?.chapter?.url?.getUrlWithoutDomain()
```
the reason this wasnt implemented before is because the chapter url used for tachi and the chapter url used for webview are different urls. For example mangadex extensions uses "api.mangadex.com" while the webpage is just "mangadex.com"

nzoba replied:
Yes that's why I did it like this: baseurl + chapterUrlWithoutDomain. I use more than 30 sources and so far every single one of them works (even mangadex) but you are right some of them might not work as expected

Jays2Kings replied:
I guess leave it for now then, someone else will bring it up on a source that doesn't work

Instead I'll ask can you add this logic to `ReaderActivity.onProvideAssistContent` as well?

Jays2Kings commented on app/src/main/java/eu/kanade/tachiyomi/smartsearch/SmartSearchEngine.kt:
```diff
@@ -127,13 +135,17 @@ class SmartSearchEngine(
      * @return a manga from the database.
      */
     suspend fun networkToLocalManga(sManga: SManga, sourceId: Long): Manga {
+        sManga.setTitleNormalized()
         var localManga = db.getManga(sManga.url, sourceId).executeAsBlocking()
         if (localManga == null) {
             val newManga = Manga.create(sManga.url, sManga.title, sourceId)
             newManga.copyFrom(sManga)
             val result = db.insertManga(newManga).executeAsBlocking()
             newManga.id = result.insertedId()
             localManga = newManga
+        } else if (localManga.title.isBlank() || localManga.title.contains("’")) {
+            localManga.title = sManga.title
+            db.insertManga(localManga).executeAsBlocking()
```
i personally rather not fix this, leave this for extensions for now

nzoba replied:
As you want, I'll revert it later. I decided to work on it after coming across [this bug](https://github.com/tachiyomiorg/tachiyomi-extensions/issues/9322)  and I did it in the app to avoid doing it for each problematic extension and since U-2019 is quite difficult to do with regular keyboard on android but yeah it might be overkill to do it here just to avoid a few cases

Jays2Kings replied:
Fixing it for migration searching is fine with me, but let's not fix it in the db, or make changes in the ui for it

Jays2Kings commented on app/src/main/java/eu/kanade/tachiyomi/util/chapter/ChapterSourceSync.kt:
```diff
@@ -128,14 +128,20 @@ fun syncChaptersWithSource(
+                    }
+                    // Try to to use the fetch date it originally had to not pollute 'Updates' tab
+                    toDelete.filter { it.chapter_number == chapter.chapter_number }
+                        .minByOrNull { it.date_fetch }!!.let {
```
you're already using let, no need for the !! that'll just lead to crashes
```suggestion
                        .minByOrNull { it.date_fetch }.let {
```

nzoba replied:
I took this feature from main tachiyomi so I didn't change much of it but yeah you are totally right, I'll replace it with a `?`

nzoba commented:
I fixed all you the things you asked for (last 2 commits) :
- Added `chapterUrl` logic to `ReaderActivity.onProvideAssistContent` + `?` in `ChapterSourceSync`
- Normalized titles only for migration search (no UI or db changes)

Let me know if any other changes are needed

Jays2Kings commented on app/src/main/java/eu/kanade/tachiyomi/smartsearch/SmartSearchEngine.kt:
```diff
@@ -54,18 +56,24 @@ class SmartSearchEngine(
 
             searchResults.mangas.map {
-                val normalizedDistance = normalizedLevenshtein.similarity(title, it.title)
+                it.setTitleNormalized()
+                val normalizedDistance = normalizedLevenshtein.similarity(titleNormalized, it.title)
```
Still, I wouldn't change the title in migration, just the search term.
```suggestion
                val normalizedDistance = normalizedLevenshtein.similarity(titleNormalized, it.title.toNormalized())
```

nzoba replied:
Done
I just added the normalization when clicking on manga title and in manual migration, tell me if you don't like it ([b100e69](https://github.com/Jays2Kings/tachiyomiJ2K/pull/1175/commits/b100e6981ced4412aeb0a78e441acf49a2852b92))

Jays2Kings commented on app/src/main/java/eu/kanade/tachiyomi/smartsearch/SmartSearchEngine.kt:
```diff
@@ -54,18 +56,24 @@ class SmartSearchEngine(
             if (searchResults.mangas.size == 1) {
-                return@supervisorScope listOf(SearchEntry(searchResults.mangas.first(), 0.0))
+                val onlyManga = searchResults.mangas.first()
+                onlyManga.setTitleNormalized()
+                return@supervisorScope listOf(SearchEntry(onlyManga, 0.0))
```
Same with this, just return the onlymanga without any conversion

Jays2Kings commented on app/src/main/java/eu/kanade/tachiyomi/source/model/SManga.kt:
```diff
@@ -77,6 +77,10 @@ interface SManga : Serializable {
             return MangaImpl()
         }
+
+        fun SManga.setTitleNormalized() {
+            this.title = this.title.replace("’", "'")
+        }
```
Yeah remove this entirely, I don't even want someone to accidentally use this

Jays2Kings commented on app/src/main/java/eu/kanade/tachiyomi/ui/library/LibraryController.kt:
```diff
@@ -1751,12 +1750,62 @@ class LibraryController(
             }
             R.id.action_mark_as_read -> {
-                presenter.markReadStatus(selectedMangas.toList(), true)
+                val markRead = true
```
im not in full love with the undo snackbar for this but eh its fine.

Instead, can you keep it, but also add a dialog for confirmation for mark as read/unread? like the one in manga details, and then show this undo snackbar.

nzoba replied:
Done
The main point of the snackbar was to allow rollback in case of unwanted click. The confirmation dialog prevents unwanted click so the snackbar is not mandatory anymore, if you really don't like it I can remove it

Jays2Kings commented on app/src/main/java/eu/kanade/tachiyomi/ui/reader/ReaderActivity.kt:
```diff
@@ -1379,13 +1380,16 @@ class ReaderActivity : BaseRxActivity<ReaderPresenter>() {
         val source = presenter.source as? HttpSource ?: return
-        val url = try {
+        val chapterUrl = presenter.getCurrentChapter()?.chapter?.url?.getUrlWithoutDomain()
+        val url = if (chapterUrl.isNullOrBlank()) try {
+            val manga = presenter.manga ?: return
             source.mangaDetailsRequest(manga).url.toString()
         } catch (e: Exception) {
+            toast(e.message)
             return
-        }
+        } else source.baseUrl + chapterUrl
```
i'd just create a helper func in presenter for building this chapter url returning a `String?`, returning if its null

No need for the toast catch either since it should never enter the catch case, the try is more of a safety measure

nzoba replied:
Done
Not sure if it was exactly what you meant but I'll let you judge ([5e013c0](https://github.com/Jays2Kings/tachiyomiJ2K/pull/1175/commits/5e013c0acb02633e691ef9d39d33a387f754a4fe))

Jays2Kings reviewed (changes requested):
Should be the final set of comments, rest of it looks good to me

Jays2Kings reviewed (approved):
Everything LGTM, I'll merge this in a bit
